File: sdks/typescript/src/services/SwiftProcessor.ts
/**
 * SWIFT Message Processing Service
 * Enhanced SwiftParser integration with blockchain routing
 */

import { LegacyBaaSClient } from '../client/LegacyBaaSClient';
import { ApiResponse } from '../types';
import { LegacyBaaSError } from '../errors';

export interface SwiftMessage {
  messageId: string;
  messageType: string;
  content: string;
  sender: string;
  receiver: string;
  timestamp: string;
  priority?: 'low' | 'normal' | 'high' | 'urgent';
  metadata?: Record<string, any>;
}

export interface SwiftProcessingOptions {
  validateOnly?: boolean;
  asyncProcessing?: boolean;
  priorityLevel?: 'low' | 'normal' | 'high' | 'urgent';
  enableBlockchain?: boolean;
  routingPreference?: 'fastest' | 'cheapest' | 'most_secure';
}

export interface SwiftProcessingResult {
  messageId: string;
  status: 'processed' | 'failed' | 'pending' | 'routing';
  blockchainTxId?: string;
  routingDecision?: {
    selectedNetwork: string;
    estimatedFees: number;
    estimatedTime: string;
    confidence: number;
  };
  validationResult: {
    isValid: boolean;
    errors: string[];
    warnings: string[];
  };
  processingTime: number;
  timestamp: string;
}

export interface MT103Data {
  transactionReference: string;
  bankOperationCode?: string;
  instructionCode?: string;
  senderToReceiverInfo?: string;
  orderingCustomer: string;
  orderingInstitution?: string;
  sendersCorrespondent?: string;
  receiversCorrespondent?: string;
  intermediaryInstitution?: string;
  accountWithInstitution?: string;
  beneficiaryCustomer: string;
  remittanceInformation?: string;
  amount: string;
  currency: string;
  valueDate: string;
}

export interface MT202Data {
  transactionReference: string;
  relatedReference?: string;
  timeIndication?: string;
  sendersCorrespondent?: string;
  receiversCorrespondent?: string;
  intermediaryInstitution?: string;
  accountWithInstitution: string;
  beneficiaryInstitution: string;
  amount: string;
  currency: string;
  valueDate: string;
}

export class SwiftProcessor {
  constructor(private client: LegacyBaaSClient) {}

  /**
   * Process SWIFT message with enhanced routing
   */
  async processMessage(
    message: SwiftMessage,
    options: SwiftProcessingOptions = {}
  ): Promise<SwiftProcessingResult> {
    const response = await this.client.makeRequest<SwiftProcessingResult>(
      'POST',
      '/swift/process',
      {
        message,
        options: {
          validateOnly: options.validateOnly || false,
          asyncProcessing: options.asyncProcessing || false,
          priorityLevel: options.priorityLevel || 'normal',
          enableBlockchain: options.enableBlockchain !== false,
          routingPreference: options.routingPreference || 'fastest'
        }
      }
    );

    return response.data;
  }

  /**
   * Process MT103 Customer Credit Transfer
   */
  async processMT103(
    mt103Data: MT103Data,
    options: SwiftProcessingOptions = {}
  ): Promise<SwiftProcessingResult> {
    const message: SwiftMessage = {
      messageId: mt103Data.transactionReference,
      messageType: 'MT103',
      content: this.formatMT103(mt103Data),
      sender: mt103Data.orderingCustomer,
      receiver: mt103Data.beneficiaryCustomer,
      timestamp: new Date().toISOString(),
      priority: options.priorityLevel || 'high'
    };

    return this.processMessage(message, {
      ...options,
      enableBlockchain: true,
      priorityLevel: 'high'
    });
  }

  /**
   * Process MT202 Financial Institution Transfer
   */
  async processMT202(
    mt202Data: MT202Data,
    options: SwiftProcessingOptions = {}
  ): Promise<SwiftProcessingResult> {
    const message: SwiftMessage = {
      messageId: mt202Data.transactionReference,
      messageType: 'MT202',
      content: this.formatMT202(mt202Data),
      sender: mt202Data.accountWithInstitution,
      receiver: mt202Data.beneficiaryInstitution,
      timestamp: new Date().toISOString(),
      priority: options.priorityLevel || 'high'
    };

    return this.processMessage(message, {
      ...options,
      enableBlockchain: true,
      priorityLevel: 'high'
    });
  }

  /**
   * Parse raw SWIFT message
   */
  async parseMessage(rawMessage: string): Promise<SwiftMessage> {
    const response = await this.client.makeRequest<SwiftMessage>(
      'POST',
      '/swift/parse',
      { rawMessage }
    );

    return response.data;
  }

  /**
   * Validate SWIFT message format
   */
  async validateMessage(message: SwiftMessage): Promise<{
    isValid: boolean;
    errors: string[];
    warnings: string[];
    messageType: string;
  }> {
    const response = await this.client.makeRequest<{
      isValid: boolean;
      errors: string[];
      warnings: string[];
      messageType: string;
    }>('POST', '/swift/validate', { message });

    return response.data;
  }

  /**
   * Get message processing status
   */
  async getMessageStatus(messageId: string): Promise<{
    messageId: string;
    status: string;
    progress: number;
    blockchainTxId?: string;
    errors?: string[];
    lastUpdate: string;
  }> {
    const response = await this.client.makeRequest<{
      messageId: string;
      status: string;
      progress: number;
      blockchainTxId?: string;
      errors?: string[];
      lastUpdate: string;
    }>('GET', `/swift/status/${messageId}`);

    return response.data;
  }

  /**
   * Get processing history
   */
  async getProcessingHistory(filters: {
    startDate?: string;
    endDate?: string;
    messageType?: string;
    status?: string;
    limit?: number;
    offset?: number;
  } = {}): Promise<{
    results: SwiftProcessingResult[];
    pagination: {
      total: number;
      limit: number;
      offset: number;
      hasMore: boolean;
    };
  }> {
    const response = await this.client.makeRequest<{
      results: SwiftProcessingResult[];
      pagination: {
        total: number;
        limit: number;
        offset: number;
        hasMore: boolean;
      };
    }>('GET', '/swift/history', undefined, {
      headers: { 'X-Query-Params': JSON.stringify(filters) }
    });

    return response.data;
  }

  /**
   * Get supported message types
   */
  async getSupportedMessageTypes(): Promise<{
    messageTypes: string[];
    capabilities: Record<string, {
      blockchainSupported: boolean;
      validationRules: string[];
      processingTime: string;
    }>;
  }> {
    const response = await this.client.makeRequest<{
      messageTypes: string[];
      capabilities: Record<string, {
        blockchainSupported: boolean;
        validationRules: string[];
        processingTime: string;
      }>;
    }>('GET', '/swift/message-types');

    return response.data;
  }

  /**
   * Get SWIFT network metrics
   */
  async getMetrics(): Promise<{
    totalMessages: number;
    successRate: number;
    averageProcessingTime: number;
    blockchainRouted: number;
    networkHealth: Record<string, number>;
  }> {
    const response = await this.client.makeRequest<{
      totalMessages: number;
      successRate: number;
      averageProcessingTime: number;
      blockchainRouted: number;
      networkHealth: Record<string, number>;
    }>('GET', '/swift/metrics');

    return response.data;
  }

  /**
   * Format MT103 message
   */
  private formatMT103(data: MT103Data): string {
    const lines = [
      `{1:${data.transactionReference}}`,
      `{2:${data.amount}${data.currency}}`,
      `{3:${data.valueDate}}`,
      `{4:`,
      `:20:${data.transactionReference}`,
      data.bankOperationCode ? `:23B:${data.bankOperationCode}` : '',
      data.instructionCode ? `:23E:${data.instructionCode}` : '',
      `:32A:${data.valueDate}${data.currency}${data.amount}`,
      `:50K:${data.orderingCustomer}`,
      data.orderingInstitution ? `:52A:${data.orderingInstitution}` : '',
      data.sendersCorrespondent ? `:53A:${data.sendersCorrespondent}` : '',
      data.receiversCorrespondent ? `:54A:${data.receiversCorrespondent}` : '',
      data.intermediaryInstitution ? `:56A:${data.intermediaryInstitution}` : '',
      data.accountWithInstitution ? `:57A:${data.accountWithInstitution}` : '',
      `:59:${data.beneficiaryCustomer}`,
      data.remittanceInformation ? `:70:${data.remittanceInformation}` : '',
      data.senderToReceiverInfo ? `:72:${data.senderToReceiverInfo}` : '',
      `-}`
    ];

    return lines.filter(line => line.trim()).join('\n');
  }

  /**
   * Format MT202 message
   */
  private formatMT202(data: MT202Data): string {
    const lines = [
      `{1:${data.transactionReference}}`,
      `{2:${data.amount}${data.currency}}`,
      `{3:${data.valueDate}}`,
      `{4:`,
      `:20:${data.transactionReference}`,
      data.relatedReference ? `:21:${data.relatedReference}` : '',
      data.timeIndication ? `:13C:${data.timeIndication}` : '',
      `:32A:${data.valueDate}${data.currency}${data.amount}`,
      data.sendersCorrespondent ? `:53A:${data.sendersCorrespondent}` : '',
      data.receiversCorrespondent ? `:54A:${data.receiversCorrespondent}` : '',
      data.intermediaryInstitution ? `:56A:${data.intermediaryInstitution}` : '',
      `:57A:${data.accountWithInstitution}`,
      `:58A:${data.beneficiaryInstitution}`,
      `-}`
    ];

    return lines.filter(line => line.trim()).join('\n');
  }
}

File: sdks/python/legacybaas/services/swift_processor.py
"""
SWIFT Message Processing Service
Enhanced SwiftParser integration with blockchain routing
"""

from typing import Dict, Any, List, Optional
from ..errors import LegacyBaaSError, SwiftProcessingError


class SwiftProcessor:
    """SWIFT Message Processor for Python SDK"""
    
    def __init__(self, client):
        self.client = client
    
    def process_message(
        self,
        message: Dict[str, Any],
        options: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Process SWIFT message with enhanced routing
        
        Args:
            message: SWIFT message data
            options: Processing options
            
        Returns:
            Processing result
        """
        default_options = {
            "validate_only": False,
            "async_processing": False,
            "priority_level": "normal",
            "enable_blockchain": True,
            "routing_preference": "fastest"
        }
        
        if options:
            default_options.update(options)
        
        try:
            return self.client._make_request(
                "POST",
                "/swift/process",
                data={
                    "message": message,
                    "options": default_options
                }
            )
        except Exception as e:
            raise SwiftProcessingError(f"SWIFT processing failed: {e}")
    
    def process_mt103(
        self,
        mt103_data: Dict[str, Any],
        options: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Process MT103 Customer Credit Transfer
        
        Args:
            mt103_data: MT103 message data
            options: Processing options
            
        Returns:
            Processing result
        """
        message = {
            "message_id": mt103_data["transaction_reference"],
            "message_type": "MT103",
            "content": self._format_mt103(mt103_data),
            "sender": mt103_data["ordering_customer"],
            "receiver": mt103_data["beneficiary_customer"],
            "timestamp": "",
            "priority": options.get("priority_level", "high") if options else "high"
        }
        
        processing_options = {
            "enable_blockchain": True,
            "priority_level": "high"
        }
        if options:
            processing_options.update(options)
        
        return self.process_message(message, processing_options)
    
    def process_mt202(
        self,
        mt202_data: Dict[str, Any],
        options: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Process MT202 Financial Institution Transfer
        
        Args:
            mt202_data: MT202 message data
            options: Processing options
            
        Returns:
            Processing result
        """
        message = {
            "message_id": mt202_data["transaction_reference"],
            "message_type": "MT202",
            "content": self._format_mt202(mt202_data),
            "sender": mt202_data["account_with_institution"],
            "receiver": mt202_data["beneficiary_institution"],
            "timestamp": "",
            "priority": options.get("priority_level", "high") if options else "high"
        }
        
        processing_options = {
            "enable_blockchain": True,
            "priority_level": "high"
        }
        if options:
            processing_options.update(options)
        
        return self.process_message(message, processing_options)
    
    def parse_message(self, raw_message: str) -> Dict[str, Any]:
        """
        Parse raw SWIFT message
        
        Args:
            raw_message: Raw SWIFT message string
            
        Returns:
            Parsed message
        """
        return self.client._make_request(
            "POST",
            "/swift/parse",
            data={"raw_message": raw_message}
        )
    
    def validate_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
        """
        Validate SWIFT message format
        
        Args:
            message: SWIFT message to validate
            
        Returns:
            Validation result
        """
        return self.client._make_request(
            "POST",
            "/swift/validate",
            data={"message": message}
        )
    
    def get_message_status(self, message_id: str) -> Dict[str, Any]:
        """
        Get message processing status
        
        Args:
            message_id: Message ID to check
            
        Returns:
            Message status
        """
        return self.client._make_request(
            "GET",
            f"/swift/status/{message_id}"
        )
    
    def get_processing_history(
        self,
        filters: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Get processing history
        
        Args:
            filters: Optional filters
            
        Returns:
            Processing history
        """
        return self.client._make_request(
            "GET",
            "/swift/history",
            params=filters or {}
        )
    
    def get_supported_message_types(self) -> Dict[str, Any]:
        """
        Get supported message types
        
        Returns:
            Supported message types and capabilities
        """
        return self.client._make_request(
            "GET",
            "/swift/message-types"
        )
    
    def get_metrics(self) -> Dict[str, Any]:
        """
        Get SWIFT network metrics
        
        Returns:
            Network metrics
        """
        return self.client._make_request(
            "GET",
            "/swift/metrics"
        )
    
    def _format_mt103(self, data: Dict[str, Any]) -> str:
        """Format MT103 message"""
        lines = [
            f"{{1:{data['transaction_reference']}}}",
            f"{{2:{data['amount']}{data['currency']}}}",
            f"{{3:{data['value_date']}}}",
            "{4:",
            f":20:{data['transaction_reference']}",
        ]
        
        if data.get("bank_operation_code"):
            lines.append(f":23B:{data['bank_operation_code']}")
        
        if data.get("instruction_code"):
            lines.append(f":23E:{data['instruction_code']}")
        
        lines.extend([
            f":32A:{data['value_date']}{data['currency']}{data['amount']}",
            f":50K:{data['ordering_customer']}",
        ])
        
        if data.get("ordering_institution"):
            lines.append(f":52A:{data['ordering_institution']}")
        
        if data.get("senders_correspondent"):
            lines.append(f":53A:{data['senders_correspondent']}")
        
        if data.get("receivers_correspondent"):
            lines.append(f":54A:{data['receivers_correspondent']}")
        
        if data.get("intermediary_institution"):
            lines.append(f":56A:{data['intermediary_institution']}")
        
        if data.get("account_with_institution"):
            lines.append(f":57A:{data['account_with_institution']}")
        
        lines.append(f":59:{data['beneficiary_customer']}")
        
        if data.get("remittance_information"):
            lines.append(f":70:{data['remittance_information']}")
        
        if data.get("sender_to_receiver_info"):
            lines.append(f":72:{data['sender_to_receiver_info']}")
        
        lines.append("-}")
        
        return "\n".join(lines)
    
    def _format_mt202(self, data: Dict[str, Any]) -> str:
        """Format MT202 message"""
        lines = [
            f"{{1:{data['transaction_reference']}}}",
            f"{{2:{data['amount']}{data['currency']}}}",
            f"{{3:{data['value_date']}}}",
            "{4:",
            f":20:{data['transaction_reference']}",
        ]
        
        if data.get("related_reference"):
            lines.append(f":21:{data['related_reference']}")
        
        if data.get("time_indication"):
            lines.append(f":13C:{data['time_indication']}")
        
        lines.extend([
            f":32A:{data['value_date']}{data['currency']}{data['amount']}",
        ])
        
        if data.get("senders_correspondent"):
            lines.append(f":53A:{data['senders_correspondent']}")
        
        if data.get("receivers_correspondent"):
            lines.append(f":54A:{data['receivers_correspondent']}")
        
        if data.get("intermediary_institution"):
            lines.append(f":56A:{data['intermediary_institution']}")
        
        lines.extend([
            f":57A:{data['account_with_institution']}",
            f":58A:{data['beneficiary_institution']}",
            "-}"
        ])
        
        return "\n".join(lines)